### PR TITLE
Feature: Add optional confirmation on save

### PR DESCRIPTION
**Pull request by G-Simpson**
I have added a conditional turbo-confirm message to the save operation on the edit page. The spinner activated by the stimulus controller needed to be updated to work without a page refresh. I have added appropriate system tests. Please let me know what you think!

Fixes https://github.com/avo-hq/avo/issues/2843

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the [documentation](https://github.com/avo-hq/avodocs)
- [x] I have added tests that prove my fix is effective or that my feature works

## Screenshots & recording
confirm_on_save

## Manual review steps

1. Step 1
1. Step 2

Manual reviewer: please leave a comment with output from the test if that's the case.

### CODE REVIEW

**Comment by binarygit** [comment]
Hey @G-Simpson Thanks for working on this!
I took a look at your changes then went source diving into the code. There might be a less verbose way of doing this, and we also need to expand this to work with (and be configurable) create|update|save actions.

There's already an if clause in `loading_button` which checks if the user needs to confirm the action or not. We should be able to pass in the appropriate value (confirmMessageValue) and have this functionality working for us.
```js
  attemptSubmit(e) {
    // If the user has to confirm the action
    if (this.confirmationMessageValue) {
      this.confirmAndApply(e)
    } else {
      this.applyLoader()
    }

    return null
  }
```
To use this, there's already code in `/avo/button_component.rb`  which is looking at the args that are getting passed and is creating the button:
```ruby
  def args
    if @args[:loading]
      @args[:"data-controller"] = "loading-button"
      @args[:"data-loading-button-confirmed-value"] = false
      @args[:"data-action"] = "click->loading-button#attemptSubmit"

      if @args[:confirm]
        @args[:"data-loading-button-confirmation-message-value"] = @args.delete(:confirm)
      end
    end

    @args[:class] = button_classes
    @args[:aria] = @aria

    @args
  end
```
Could you look at how we can pass the `args[:confirm]` arg into our SaveButton component?

**Comment by G-Simpson** [comment]
@binarygit thanks for getting back to me with a detailed response.

I've done a bit of digging and it seems that the `this.confirmAndApply(e)` is redundant. When I specify a turbo_confirm [here](https://github.com/avo-hq/avo/blob/main/app/components/avo/resource_component.rb#L185), in a way that's similar to the delete button, we don't trigger the condition you mentioned above [here](https://github.com/avo-hq/avo/blob/main/app/components/avo/button_component.rb#L33).  If I specifically add a `confirm` option, then we do trigger it, although it is the standard js confirm, not the one that has been [hijacked by turbo](https://github.com/avo-hq/avo/blob/main/app/javascript/js/custom-confirm.js) which I assume is the one we want going forward.

An example:
js-confirm

I would like to remove that redundant code but I'm not sure how the maintainers of this project feel about whether that should be part of this PR or another. I tilt towards another but am happy either way.

I agree that where I defined the original turbo_confirm on the form wasn't the best place for it, so I've moved it to match the pattern of the other button.

File: app/javascript/js/controllers/loading_button_controller.js [hunk 1]
```diff
@@ -11,6 +11,16 @@ export default class extends Controller {
...
+  connect() {
+    this.button.setAttribute('data-original-content', this.button.innerHTML)
+    this.dialog = document.getElementById('turbo-confirm')
+    this.dialogCloseHandler = this.handleDialogClose.bind(this)
+
+    this.dialog.addEventListener('close', this.dialogCloseHandler)
+  }
+  disconnect() {
+    this.dialog.removeEventListener('close', this.dialogCloseHandler)
+  }
```
Comment (Paul-Bob) [inline app/javascript/js/controllers/loading_button_controller.js:23]:
```suggestion

  connect() {
    this.button.setAttribute('data-original-content', this.button.innerHTML)
    this.dialog = document.getElementById('turbo-confirm')
    this.dialogCloseHandler = this.handleDialogClose.bind(this)

    this.dialog.addEventListener('close', this.dialogCloseHandler)
  }

  disconnect() {
    this.dialog.removeEventListener('close', this.dialogCloseHandler)
  }
```

**Review by Paul-Bob — COMMENTED** [review]
Hei @G-Simpson thanks for digging into this.

I let some lint suggestions

I also checked the code regrading the `confirm` and `attemptSubmit` discussion with @binarygit . It seems that `attemptSubmit` is indeed redundant and not used anymore

We can follow 2 approaches

1 - remove the redundant code from the button component and from the JS controller 
2 - keep that code and make it work with the custom dialog

For both approaches we need to ensure that the correct behavior will be applied for all buttons that have a combination of `loading: true` and a `confirm` message

**Comment by G-Simpson** [comment]
@Paul-Bob I'll remove it this PR, thanks

**Comment by G-Simpson** [comment]
The only button with `loading: true` was that button, I've removed the redundant code. all forms of button confirmation now work the same way (`turbo_confirm`)

**Comment by Paul-Bob** [comment]
> The only button with `loading: true` was that button, I've removed the redundant code. all forms of button confirmation now work the same way (`turbo_confirm`)

I've also checked our private repos for `loading: true`; it was indeed used only on the save button.

**Comment by Paul-Bob** [comment]
@G-Simpson I just merged a PR that changes all the code from `Avo::BaseResource` to `Avo::Resources::Base` Can you please solve the conflicts?

**Review by Paul-Bob — APPROVED** [review]
Thank you for this contribution @G-Simpson! It's looking good and working as expected.

**Pull request merged**